### DISCUSSION TITLE
Best practices around element IDs

### DISCUSSION THREAD

**Discussion by mattie47** [zauberzeug/nicegui#3842] [Q&A]
### Question

Hi,

I spent some time today trying to debug an issue around not knowing where a certain DIV was coming from.

Whilst the situation was quite unique (I'll elaborate below), it left me wanting to be more explicit with declaration of elements and having identifiers for the element.

**To my main question - What would best practise be for adding IDs to elements we add with NiceGui?**

For example, you can do something like:

```python
#!/usr/bin/env python3
from nicegui import ui

with ui.column().style('row-gap: 0.1rem').props('label=column_container id=column_container') as column_container:
    with ui.row().style('align-items: center; width: 100%;').classes('w-full no-wrap q-py-none'):
        ui.label("THING1").style('flex: 6; font-weight: bold; font-size: 24px').props('label=thing1 id=thing1_var myprojectID=thing1_var')
        ui.label("THING2").style('flex: 6; font-weight: bold; font-size: 24px').props('label=thing2 id=thing2_var')

    with ui.row().style('align-items: center; width: 100%; margin-top: 0').classes('w-full no-wrap'):
        ui.label("Thing Info").style('flex: 12; font-size: 20px')
        ui.label("Thing Info").style('flex: 12; font-size: 20px')
ui.run()
```

which renders:
![image](https://github.com/user-attachments/assets/69e5877c-3335-46d0-b759-291065cc3e67)

This of course overwrites the dynamically generated div/element id created by default for niceGui, but it does mean if you're stuck trying to figure out where an id is being created, you don't have to add a bunch of print statements for each element with element.id to see what gets generated. 

**If we do decide to add an identifier to objects of some respect, what's the best practise around this?**
From the above, we can:
* Just update "id"
* use a "label", 
* simply introduce our own unique thing like "myprojectid=foo" ? 

I don't have a web dev background, so rather than simply pick one and go with it, I thought I'd ask here what's recommended?  :) 

Thanks!

-------------

**_For today's problem:_**

I had an empty div appearing on the page as the first item in every page. This was causing unexpected extra padding at the start of each page's content div:
```html
<div id="c3" class="nicegui-content">
  <div_id="c4"> **<--**
```

Without sharing my code, I could see what was generating ids c1, c2, c3, and c5 onwards, but not c4.

For c1, c2, and c3, I could see they were generated by `nicegui/client.py`
```python

        with Element('q-layout', _client=self).props('view="hhh lpr fff"').classes('nicegui-layout') as self.layout:
            with Element('q-page-container') as self.page_container:
                print(f'container.id= {self.page_container.id}')
                with Element('q-page'):
                    self.content = Element('div').classes('nicegui-content')
                    print(f' self.content.id = {self.content.id}')
```

which presented:
```
user@HOST~ $python main.py -m
container.id= 1
self.content.id = 3
```

I tried adding debug to my code such as the following to try figure out where c4 was coming from:

```python
for element in context.client.elements.values():
            if element.id == 4:
                element_list = list(element.__dict__.items())
                print(element_list)
                ancestors_list = list(element.ancestors())
                print("Ancestors: ", ancestors_list)
                element.ancestors
                print(element.id, element.tag, element.props, element.ancestors(), element.component, element.tag, element.classes(), element.style(), element.descendants())
```

However this didn't get me much further and I started simply commenting out large swaths of code. 

It turns out someone else working on our project had added:
```python
ui_elem = ui.element()
```

To one of the sub-pages, that was getting activated as part of an import statement at the start of our main file. 

```python
from gui.components.pi_page import createPiTab
```

So I was glad to figure this out, and thought the process of troubleshooting I went through might be useful for someone else here :)

**Comment by falkoschindler** [comment]
Hi @mattie47, thanks for sharing your experience!

NiceGUI's element IDs are a bit of a pain point. In Python they are integers, but because HTML doesn't like them to be pure numbers, we add a "c" prefix for "component". There have been proposals to add a customizable ID attribute (e.g. #1583), but so far we didn't see a good use case for it and we haven't been convinced that the user should be able to modify these internals. Even worse, there is no guarantee that there will be an element in the DOM with the corresponding ID. Take `ui.input` for example: It gets translated into a `<label>` tag with several nested elements, none of which has an ID.

For debugging purposes, like in your case, you can simply add something like `.classes('TEST')` or `.props('foo=bar')` to find the element in the DOM. In web development it is also rather common to use the [`data` attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) for extra information.
